companion.py

```diff
@@ -545,176 +694,356 @@ def auth_callback(self) -> None:
-            # this be an empty list when you've jumped to another system
-            # and not yet docked.  As opposed to no services key at all
-            # or an empty dict.
-            logger.error(f'services is "{type(services)}", not dict !')
-            if __debug__:
-                self.dump_capi_data(data)
-
-            # Set an empty dict so as to not have to retest below.
-            services = {}
-
-        last_starport_id = int(last_starport.get('id'))
-
-        if services.get('commodities'):
-            marketdata = self.query(URL_MARKET)
-            if last_starport_id != int(marketdata['id']):
-                logger.warning(f"{last_starport_id!r} != {int(marketdata['id'])!r}")
-                raise ServerLagging()
+        Perform a query against the specified CAPI endpoint.
+
+        :param endpoint: The CAPI endpoint to query.
+        :param tk_response_event: Name of tk event to generate when response queued.
+        :param query_time: When this query was initiated.
+        :param retrying: Whether this is a retry.
+        :param play_sound: Whether the app should play a sound on error.
+        :param auto_update: Whether this request was triggered automatically.
+        """
+        logger.trace_if('capi.query', f'Performing query for endpoint "{endpoint}"')
+        if self.state == Session.STATE_INIT:
+            if self.login():
+                self.query(
+                    endpoint, query_time, tk_response_event=tk_response_event, play_sound=play_sound,
+                    auto_update=auto_update
+                )
+                return
 
-            else:
-                marketdata['name'] = last_starport_name
-                data['lastStarport'].update(marketdata)
+        elif self.state == Session.STATE_AUTH:
+            logger.error('cannot make a query when unauthorized')
+            raise CredentialsError('cannot make a query when unauthorized')
 
-        if services.get('outfitting') or services.get('shipyard'):
-            shipdata = self.query(URL_SHIPYARD)
-            if last_starport_id != int(shipdata['id']):
-                logger.warning(f"{last_starport_id!r} != {int(shipdata['id'])!r}")
-                raise ServerLagging()
+        logger.trace_if('capi.query', 'Trying...')
+        if conf_module.capi_pretend_down:
+            raise ServerConnectionError(f'Pretending CAPI down: {endpoint}')
 
-            else:
-                shipdata['name'] = last_starport_name
-                data['lastStarport'].update(shipdata)
-# WORKAROUND END
+        self.capi_query_queue.put(
+            EDMCCAPIRequest(
+                endpoint=endpoint,
+                tk_response_event=tk_response_event,
+                retrying=retrying,
+                query_time=query_time,
+                play_sound=play_sound,
+                auto_update=auto_update
+            )
+        )
+
+    def profile(
+            self,
+            query_time: int = int(time.time()),
```

Comment:
This does not work as you expect. Its evaluated once at class instantiation time.

```
In [1]: from time import time

In [2]: class Test:
   ...:     def print_time(self, t: int = int(time())):
   ...:         print(t)
   ...:

In [3]: x = Test()

In [4]: x.print_time()
1629847121

In [5]: x.print_time()
1629847121
```

you will want to use an `Optional[int]` and set it within the method

Reply:
Changed ... although we might end up removing the whole function.  It's only still there 'for' stats.py, and I intend to change that to use cached data anyway.